dashboard/app/controllers/projects_controller.rb

```diff
@@ -310,26 +310,55 @@ def create_new
     )
   end
 
-  # GET /projects(/script/:script_id)/level/:level_id
+  # GET /projects(/script/:script_id)/level/:level_id(/user/:user_id)
   # Given a level_id and the current user (or signed out user), get the existing project
   # or create a new project for that level and user. If a script_id is provided, get or
-  # create the project for that level, script and user
+  # create the project for that level, script and user.
+  # If a user_id is provided, get but do not create the project for that level, script and
+  # user_id; this is used for Lab2 levels when a teacher views a student's work, and the
+  # relevant permission is verified here.
   # Returns json: {channel: <encrypted-channel-token>}
   def get_or_create_for_level
     script_id = params[:script_id]
     level = Level.find(params[:level_id])
+    user_id = params[:user_id]
+
     error_message = under_13_without_tos_teacher?(level)
     return render(status: :forbidden, json: {error: error_message}) if error_message
-    # get_storage_id works for signed out users as well, it uses the cookie to determine
-    # the storage id.
-    user_storage_id = get_storage_id
-    # Find the channel for the user and level if it exists, or create a new one.
-    channel_token = ChannelToken.find_or_create_channel_token(level, request.ip, user_storage_id, script_id, {hidden: true})
-    script_name = !script_id.nil? && Unit.find(script_id)&.name
+
+    if user_id
+      # If viewing another user's work, ensure that we have permission.
+      script_level = level.script_levels.find_by_script_id(script_id)
+      user = User.find(user_id)
+      unless can?(:view_as_user, script_level, user)
+        return render(status: :forbidden, json: {error: "Access denied."})
```

Comment:
Just flagging this for later, but do we need to handle this in any special way on the client? I'm curious what we do for other labs